lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -577,26 +606,41 @@ monitorMetadata tr gp fetchMetadata DBLayer{..} = forever $ do
         f = unActiveSlotCoefficient (getActiveSlotCoefficient gp)
 
 data StakePoolLog
-    = MsgFollow FollowLog
+    = MsgDb PoolDbLog
```

Comment:
We already have a logger for the pool database, why is this needed here?

Reply:
@KtorZ 
> We already have a logger for the pool database, why is this needed here?

Good question.

When we call the `removeRetiredPools` function, it expects to be provided with a `Tracer IO PoolDbLog` value. But we call it from a context where we only have a `Tracer IO StakePoolLog`. Therefore, we need to have some way to transform the `Tracer IO StakePoolLog` into a `Tracer IO PoolDbLog`.

Luckily, the `contramap` function allows us to perform this transformation, but it needs to be provided with something that maps a `PoolDbLog` into a `StakePoolLog`, which is where the `MsgDb` constructor comes in.

Reply:
That seems quite wrong to me. Why isn't `removeRetiredPools` part of the DBLayer already? This creates an inconsistency with the "normal approach" which seems unjustified to me. 

The problem with having logs database embedded in the wrong logger is that it completely defies the purpose of having separate logger components. Now, if I run `--trace-pool-db off`, I still get pool db logs regarding the garbage collection of pools.

Reply:
> Why isn't `removeRetiredPools` part of the DBLayer already?

That's a good question.

The `removeRetiredPools` function depends **only** on the DBLayer **interface**, and not any particular implementation. It only calls `listRetiredPools` and `removePools`. So this function can sit outside of the `DBLayer` interface itself (in the same way that we do for other function that only depend on the interface).

One alternative would be to add a `removeRetiredPools` function to the `DBLayer` type, and then implement this twice (once in the `SQLite` module and another time in the `MVar` module), but reuse the same function behind the scenes.

Reply:
> One alternative would be to add a removeRetiredPools function to the DBLayer type, and then implement this twice (once in the SQLite module and another time in the MVar module), but reuse the same function behind the scenes.

Yes. That sounds preferable for the consumers of the DBLayer such that things like the DBLog doesn't leak outside of the DBLayer.

Reply:
@KtorZ wrote:
> The problem with having logs database embedded in the wrong logger is that it completely defies the purpose of having separate logger components. Now, if I run --trace-pool-db off, I still get pool db logs regarding the garbage collection of pools.

The pattern I've used for `MsgDb` is also used by `MsgFollow`, which embeds entries from the `FollowLog`:

```hs
data StakePoolLog 
    = MsgDb PoolDbLog 
    | MsgFollow FollowLog
    ...

data FollowLog
    = MsgFollowAction (FollowAction String)   
    | MsgUnhandledException Text  
    | MsgNextBlockFailed ErrGetBlock 
    ....
```

See: https://github.com/input-output-hk/cardano-wallet/blob/9f93d0a98be7c19035dd603149ebf06ac38d7f2f/lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs#L484

Is there a better solution that we can use here?

Reply:
@jonathanknowles FollowLog is a completely different story, it is not part of any other logger component. There are two components that include a `FollowLog` and these are the `stake-pool-engine` and `wallet-engine` loggers. 

However, the DBLog is already a logger component all by itself. See:

https://github.com/input-output-hk/cardano-wallet/blob/master/lib/shelley/src/Cardano/Wallet/Shelley.hs#L463-L472

Reply:
@KtorZ many thanks for your explanation, that makes sense.

I've made the following changes:
* moved `removeRetiredPools` into the `DBLayer`
* completely removed the `MsgDb` constructor.

See the following commits:
- 2ed42a2e0b1ba8b9a939a855c6ee40505fd348ff
- 742027cb266cb77e6f7bfd073211bbb84bb404d1